### PR TITLE
Zk Tx Generator

### PR DESCRIPTION
**Pull request by sifnoc**
- Can testing with Multiple Wallet
  `docker-compose -f docker-compose.localtest.yml up --scale wallet=5`
- Each generator recognize container name then using index number as the account
- Generator can handle only one kind asset(eth for now)

### CODE REVIEW

**Comment by wanseob** [comment]
@sifnoc would you please rebase this branch to the current develop branch and make them pass the ci testing?

**Comment by wanseob** [comment]
@sifnoc looks good to me, let's merge this branch to `develop`

**Comment by sifnoc** [comment]
How are you think that changing the name of the package as `stress-test` package for now until leaving this repo?
I still believe that is a good idea to create a new repo for the generator package. but this package needs a little bit more development for a new place.
 Also thinking about the organizer API might help the operating coordinator easily access layer1 and layer2 data.

**Comment by JChanceHud** [comment]
My main issue is adding the files outside of the generator package (`genesis.json`, various docker files, kube configs, etc) and adding new dependencies to the whole system. The docker files are already a bit overwhelming. Looking at the diffs it seems like the package should pretty cleanly be split, almost all the files are contained inside the generator folder.

I know it's a pain to split the package to a separate repo after building inside this repo but I think it will make it much easier to setup launch scripts, config instructions, and to run against new versions of zkopru (as well as to update the generator).

The organizer api looks super useful! I think it would be best used in parallel to the zkopru api. For example `localhost:8888` is zkopru and `localhost:8889` is organizer. This way the organizer can be less concerned with security/exposing sensitive info.

I'm happy to help migrate/setup in another repo, it will help me understand the system better.

@wanseob Should we begin publishing the Zkopru packages to npm at some point?

**Comment by sifnoc** [comment]
> The docker files are already a bit overwhelming. Looking at the diffs it seems like the package should pretty cleanly be split, almost all the files are contained inside the generator folder.

I agree that the docker files are not a simple thing for staying in this repo now. and Thank you for noticing that the package is enough to be split, again.

> I know it's a pain to split the package to a separate repo after building inside this repo but I think it will make it much easier to setup launch scripts, config instructions, and to run against new versions of zkopru (as well as to update the generator).

Ouch! you totally read my mind. I was worried about my tasks are behind the schedule. So I have wanted to be the package stay a bit here just before starting the stress test at a cluster level. but now, I am encouraged by your comments!

> I'm happy to help migrate/setup in another repo, it will help me understand the system better.

I am so glad to hear that. how about the close this PR for the next step?
Then I can make a new PR for an organizer by your suggestion that exposes beside with the coordinator API.

**Comment by JChanceHud** [comment]
Hey @sifnoc I changed some things and created [this](https://github.com/zkopru-network/stress-test) repo. If you clone and run `sh scripts/setup.sh` it should be semi-usable.

There's currently a problem with building zkopru from an outside Typescript project, which is a good problem to solve.

I'd really rather get it working externally instead of merging and then untangling. If I can't get it working tomorrow we can merge here and not burn too much time on it.

**Comment by wanseob** [comment]
> @wanseob Should we begin publishing the Zkopru packages to npm at some point?

yes after we fix the utxo root bug & final audit verification let's release them to npm

**Comment by JChanceHud** [comment]
Hey guys, I have the external repo working now.

You can test it by running the following (to try it in a temp dir):
```sh
cd /tmp
git clone https://github.com/zkopru-network/stress-test.git
cd stress-test
sh scripts/setup.sh
yarn
yarn build
docker-compose -f compose/docker-compose.localtest-geth.yml build
docker-compose -p zkopru -f compose/docker-compose.localtest-geth.yml up --force-recreate
```

It clones the zkopru repo into the generator dir and then sets up node modules from there. Once the modules are published to npm this will not be necessary because we can use yarn in the docker image to pull/setup.

File: packages/contracts/package.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
     "pregenTS": "shx mkdir -p dist",
     "precommit": "#lint-staged",
     "compile": "truffle compile",
-    "build:contract": "yarn compile && yarn genTS && tsc --build tsconfig.build.json && yarn prettier --write src/**/*",
+    "build:contract": "yarn compile && yarn genTS && tsc --build tsconfig.build.json && yarn prettier --write src/**/* && shx cp -r src/contracts dist",
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/contracts/package.json:40]: > 1. add zkopru-network/zkopru as a submodule into stress-test repo

Yeah, That is my next step after this PR merged! Thanks.

> Move genesis.json & run_geth.sh file (or other compose files if needed) to `stress-test` repo

I forgot to remove these files in `zkopru/packages/contract`. Chance already moved the files to `stress-test` repo.

[the remark this answers is not in the document]
Comment (wanseob) [reply packages/contracts/package.json:40]: thx @sifnoc how do you think about the comment about the `dist` folder commands?

[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/contracts/package.json:40]: > dist directory is for the distribution and we're not planning to include solidity files inside the dist folder.

`stress-test` could not build without this, for now. I am trying to find another way to build without this copy.

[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/contracts/package.json:40]: > thx @sifnoc how do you think about the comment about the dist folder commands?

I just realized that the `dist` folders are using for running containers. This is a simple way to build `stress-test` but it makes unnecessary files are included in a container image.

I am still working on this, to build `stress-test` without this. :)

File: packages/babyjubjub/package.json [hunk 2]
```diff
@@ -12,7 +12,7 @@
   },
   "scripts": {
     "prebuild": "shx mkdir -p dist",
-    "build": "tsc --build tsconfig.build.json",
+    "build": "tsc --build tsconfig.build.json && shx cp -r src/types dist",
```
[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/babyjubjub/package.json]: There is a problem with finding type when it builds in `stress-test` repo.

Can be replaced with cc41a24.

File: packages/contracts/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/contracts/package.json]: Found and updated at 938f0e0

File: packages/babyjubjub/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/babyjubjub/package.json]: > types definitions in the src/types should be automatically part of the dist output without copying files. Was there a problem with this command?

I wrong answered your question. Yes, typescript declaration-type files are not copied while typescript compiling. so there's no `src/types/*.d.ts` in `contracts/dist`. 

One simple idea is that just changed `*.d.ts` to `.ts`. It seems working for me.

I was so much spent time searching for another way to solve this problem, but I do not know if it's worth the headache.

[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/babyjubjub/package.json]: Changed file extension name at c9f0908

File: packages/contracts/src/layer1.ts [hunk 3]
```diff
@@ -1,24 +1,24 @@
...
+import { IHeaderValidator } from '../src/contracts/IHeaderValidator'
+import { IMigratable } from '../src/contracts/IMigratable'
+import { IMigrationValidator } from '../src/contracts/IMigrationValidator'
+import { INullifierTreeValidator } from '../src/contracts/INullifierTreeValidator'
+import { ISetupWizard } from '../src/contracts/ISetupWizard'
+import { ITxValidator } from '../src/contracts/ITxValidator'
+import { IUserInteractable } from '../src/contracts/IUserInteractable'
+import { IUtxoTreeValidator } from '../src/contracts/IUtxoTreeValidator'
+import { IWithdrawalTreeValidator } from '../src/contracts/IWithdrawalTreeValidator'
+import { Zkopru } from '../src/contracts/Zkopru'
```
Comment (wanseob) [inline packages/contracts/src/layer1.ts:21]: let's revert these changes ;)

File: packages/contracts/src/zkopru.ts [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
+import { IHeaderValidator } from '../src/contracts/IHeaderValidator'
+import { IMigratable } from '../src/contracts/IMigratable'
+import { IMigrationValidator } from '../src/contracts/IMigrationValidator'
+import { INullifierTreeValidator } from '../src/contracts/INullifierTreeValidator'
+import { ISetupWizard } from '../src/contracts/ISetupWizard'
+import { ITxValidator } from '../src/contracts/ITxValidator'
+import { IUserInteractable } from '../src/contracts/IUserInteractable'
+import { IUtxoTreeValidator } from '../src/contracts/IUtxoTreeValidator'
+import { IWithdrawalTreeValidator } from '../src/contracts/IWithdrawalTreeValidator'
+import { Zkopru } from '../src/contracts/Zkopru'
```
Comment (wanseob) [inline packages/contracts/src/zkopru.ts:15]: let's revert these changes

File: packages/contracts/src/tx-util.ts [hunk 5]
```diff
@@ -8,7 +8,7 @@ import {
   NonPayableTransactionObject,
   PayableTx,
   NonPayableTx,
-} from './contracts/types'
+} from '../src/contracts/types'
```
Comment (wanseob) [inline packages/contracts/src/tx-util.ts:11]: let's revert this line ;)

**Review by wanseob — CHANGES_REQUESTED** [review]

File: packages/contracts/src/zkopru.ts (hunk 4 above)
Comment (sifnoc) [reply packages/contracts/src/zkopru.ts]: Updated at 33b23ae

File: packages/contracts/src/tx-util.ts (hunk 5 above)
Comment (sifnoc) [reply packages/contracts/src/tx-util.ts:11]: Updated at 33b23ae

File: packages/contracts/src/layer1.ts (hunk 3 above)
Comment (sifnoc) [reply packages/contracts/src/layer1.ts]: Updated 33b23ae

File: packages/contracts/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (sifnoc) [reply packages/contracts/package.json]: I tried to convert `contracts/src/contracts/*.d.ts` to `.ts` with newly added script in utlis. But it makes more time to build in `zkopru` repo. So It would be better to treat this in `stress-test` repo without using the copy command or something similar in `zkopru` repo.

File: packages/contracts/src/tx-util.ts (hunk 5 above)
Comment (wanseob) [reply packages/contracts/src/tx-util.ts:11]: Maybe that commit is missing this file? would u plz check this once again?

File: packages/zk-wizard/src/zk-wallet-account.ts [hunk 6]
```diff
@@ -728,6 +734,7 @@ export class ZkWalletAccount {
         .eth()
         .add(fee)
         .toString(),
+      gas: 400000, // TODO : minimum gas set inside 'sendTx' with function signature
```
Comment (wanseob) [inline packages/zk-wizard/src/zk-wallet-account.ts:737]: Does the TODO comment mean this logic?
https://github.com/zkopru-network/zkopru/blob/14e0d5147c7e2868b69cb51cf850a95852358615/packages/contracts/src/tx-util.ts#L19-L58

**Comment by wanseob** [comment]
1. The convention for type declaration files is using .d.ts for its extension name. So files in the "types" directory should maintain their extension names for the common convention. If you are having problem in referencing the types outside of the repository for example in the "stress-test" you may need to figure tsconfig.json of that repository.
2. After all the changes please allow me to merge this PR w/ the squash & merge option :)

**Comment by sifnoc** [comment]
> 1. The convention for type declaration files is using .d.ts for its extension name. So files in the "types" directory should maintain their extension names for the common convention. If you are having problem referencing the types outside of the repository for example in the "stress-test" you may need to figure tsconfig.json of that repository.

I see, agree with the convention in `types` directory. Is it okay with changing the name `d.ts` to `ts` in c9f0908?
I could not find a way to point out importing path in `.d.ts` differently. Of course, can do a common `ts` file with `tsconfig.json`, but I do not know how to handle `.d.ts` with tsconfig. If you have an idea about it, please let me.

FYI, these compile errors in `stress-test` repo.

```ts
node_modules/@zkopru/contracts/dist/layer1.d.ts:20:24 - error TS2307: Cannot find module './contracts/Zkopru' or its corresponding type declarations.

20 import { Zkopru } from './contracts/Zkopru'
                          ~~~~~~~~~~~~~~~~~~~~

node_modules/@zkopru/contracts/dist/tx-util.d.ts:8:8 - error TS2307: Cannot find module './contracts/types' or its corresponding type declarations.

8 } from './contracts/types'
         ~~~~~~~~~~~~~~~~~~~
``` 

I will be happy if I know this
> 2. After all the changes please allow me to merge this PR w/ the squash & merge option :)

Of course! It too messy to merge directly! :)

File: packages/contracts/src/tx-util.ts (hunk 5 above)
Comment (sifnoc) [reply packages/contracts/src/tx-util.ts]: Forgot to revert this one.
resolved c5e2c52

File: packages/zk-wizard/src/zk-wallet-account.ts (hunk 6 above)
Comment (sifnoc) [reply packages/zk-wizard/src/zk-wallet-account.ts]: lost my comment!

shortly said that I forgot to remove this one. ;)
And 'TODO' means that I want the gas amount is following a 'gasTable' but it is not created yet. 

Removed at c5e2c52

**Comment by sifnoc** [comment]
@wanseob @JChanceHud Let me merge this PR :)

**Comment by wanseob** [comment]
@sifnoc good please merge this PR w/ "Squash and merge" option :)

**Review by wanseob — CHANGES_REQUESTED** [review]
oops let's make the final change. yarn.lock is including unnecessary packages like redis bullmq and etc. Would u please revert the changes in the yarn.lock?

**Review by JChanceHud — APPROVED** [review]
